### PR TITLE
Address index

### PR DESCRIPTION
**Pull request by raphjaph**
This is a very crude address index, to make this production ready we have to either add an `OUTPOINT_TO_TXOUT` table, or adapt the value cache and fetcher to also get the script pubkey in addition to the output value.

Some more considerations:

- At the moment the table stores the bytes of the script_pubkey as the key. Probably smarter to use the hash of the script_pubkey so that it is of fixed size. Look up would then go Address -> ScriptPukey -> Hash(ScriptPubkey): `SCRIPT_PUBKEY_HASH_TO_OUTPOINT`
- Reuse the `TRANSACTION_ID_TO_TRANSACTION` table.

### CODE REVIEW

File: src/index.rs [hunk 1]
```diff
@@ -501,6 +518,7 @@ impl Index {
     content_type_counts.sort_by_key(|(_content_type, count)| Reverse(*count));
 
     Ok(StatusHtml {
+      address_index: statistic(Statistic::IndexAddresses)? != 0,
```
Comment (raphjaph) [inline src/index.rs:521]: use self.has_address_index() and refactor others as well

File: src/index.rs [hunk 2]
```diff
@@ -1636,10 +1654,20 @@ impl Index {
     Ok(
       outpoint != OutPoint::null()
         && outpoint != self.settings.chain().genesis_coinbase_outpoint()
-        && self
-          .client
-          .get_tx_out(&outpoint.txid, outpoint.vout, Some(true))?
-          .is_none(),
+        // If we are not indexing inscriptions or addresses this table does not get built
+        && if self.settings.index_inscriptions() || self.settings.index_addresses() {
```
Comment (raphjaph) [inline src/index.rs:1658]:
```suggestion
        && if self.settings.index_addresses() {
```

File: src/index.rs [hunk 3]
```diff
@@ -6268,6 +6310,70 @@ mod tests {
...
+      Vec::new(),
+    );
+
+    assert_eq!(
+      context.index.get_address_info(&second_address).unwrap(),
+      vec![OutPoint {
+        txid: transaction.txid(),
+        vout: 0
+      }]
+    );
```
Comment (raphjaph) [inline src/index.rs:6374]: do with slices

File: src/index.rs [hunk 4]
```diff
@@ -6268,6 +6310,70 @@ mod tests {
...
+    let second_address = context
+      .index
+      .settings
+      .chain()
+      .address_from_script(&transaction.output[0].script_pubkey)
+      .unwrap();
+
+    assert_eq!(
+      context.index.get_address_info(&first_address).unwrap(),
+      Vec::new(),
```
Comment (raphjaph) [inline src/index.rs:6365]:
```suggestion
      [],
```

File: src/index.rs [hunk 5]
```diff
@@ -6268,6 +6310,70 @@ mod tests {
...
+      .chain()
+      .address_from_script(&transaction.output[0].script_pubkey)
+      .unwrap();
+
+    assert_eq!(
+      context.index.get_address_info(&first_address).unwrap(),
+      vec![OutPoint {
+        txid: transaction.txid(),
+        vout: 0
+      }]
```
Comment (raphjaph) [inline src/index.rs:6343]:
```suggestion
      [OutPoint {
        txid: transaction.txid(),
        vout: 0
      }]
```

File: src/index.rs [hunk 6]
```diff
@@ -6268,6 +6310,70 @@ mod tests {
       .unwrap());
   }
 
+  #[test]
+  fn output_addresses_are_updated() {
```
Comment (raphjaph) [inline src/index.rs:6314]: Add a test for when an address has two outpoints

File: src/index/entry.rs [hunk 7]
```diff
@@ -484,6 +504,19 @@ impl Entry for Txid {
 mod tests {
   use super::*;
 
+  #[test]
+  fn txout_entry() {
+    let txout = TxOut {
+      value: u64::MAX,
+      script_pubkey: ScriptBuf::new(),
```
Comment (raphjaph) [inline src/index/entry.rs:511]: put an actual address in here

File: src/index/updater.rs [hunk 8]
```diff
@@ -325,51 +346,79 @@ impl<'index> Updater<'index> {
...
+          // Send this outpoint to background thread to be fetched
+          output_sender.blocking_send(prev_output)?;
         }
       }
     }
 
+    if self.index.index_addresses {
+      let Some(address_txout_receiver) = address_txout_receiver else {
+        unreachable!()
+      };
```
Comment (raphjaph) [inline src/index/updater.rs:407]:
```suggestion
    if let Some(address_txout_receiver) = address_txout_receiver else {
        unreachable!()
      };
```

File: src/index/updater.rs [hunk 9]
```diff
@@ -630,6 +679,60 @@ impl<'index> Updater<'index> {
     Ok(())
   }
 
+  fn index_transaction_addresses(
```
Comment (raphjaph) [inline src/index/updater.rs:682]:
```suggestion
  fn index_transaction_outputs(
```

File: src/index/updater.rs [hunk 10]
```diff
@@ -630,6 +679,60 @@ impl<'index> Updater<'index> {
...
+        continue;
+      }
+
+      // multi-level cache for UTXO set to get to the script pubkey
+      let txout = if let Some(txout) = utxo_cache.get(&txin.previous_output) {
+        txout.clone()
+      } else if let Some(value) = outpoint_to_txout.get(&txin.previous_output.store())? {
+        TxOut::load(value.value())
+      } else {
+        txout_receiver.blocking_recv().map_err(|_| {
```
Comment (raphjaph) [inline src/index/updater.rs:704]: maybe add error message from channel in here
```suggestion
        txout_receiver.blocking_recv().map_err(|err| {
```

File: src/index/updater.rs [hunk 11]
```diff
@@ -630,6 +679,60 @@ impl<'index> Updater<'index> {
...
+          anyhow!(
+            "failed to get transaction for {}",
+            txin.previous_output.txid
+          )
+        })?
+      };
+
+      // We remove these values in the InscriptionUpdater only when indexing inscriptions
```
Comment (raphjaph) [inline src/index/updater.rs:712]:
```suggestion
      // If we are indexing inscriptions, the InscriptionUpdater will remove these.
```

File: src/index/updater.rs [hunk 12]
```diff
@@ -630,6 +679,60 @@ impl<'index> Updater<'index> {
...
+      let vout: u32 = vout.try_into().unwrap();
+      script_pubkey_to_outpoint.insert(
+        txout.script_pubkey.as_bytes(),
+        OutPoint { txid: *txid, vout }.store(),
+      )?;
+
+      utxo_cache
+        .entry(OutPoint { txid: *txid, vout })
+        .or_insert(txout.clone());
+    }
```
Comment (raphjaph) [inline src/index/updater.rs:731]:
```suggestion
      utxo_cache
        .insert(OutPoint { txid: *txid, vout }, (txout.clone());
    }
```

File: src/index/updater/inscription_updater.rs [hunk 13]
```diff
@@ -114,24 +114,23 @@ impl<'a, 'tx> InscriptionUpdater<'a, 'tx> {
...
-        .outpoint_to_value
-        .remove(&tx_in.previous_output.store())?
+        .outpoint_to_txout
+        .remove(&txin.previous_output.store())?
       {
-        value.value()
+        TxOut::load(value.value())
       } else {
-        self.value_receiver.blocking_recv().ok_or_else(|| {
+        self.txout_receiver.blocking_recv().map_err(|_| {
```
Comment (raphjaph) [inline src/index/updater/inscription_updater.rs:125]:
```suggestion
        self.txout_receiver.blocking_recv().map_err(|err| {
```

File: src/index/updater/inscription_updater.rs [hunk 14]
```diff
@@ -314,13 +313,13 @@ impl<'a, 'tx> InscriptionUpdater<'a, 'tx> {
...
-        OutPoint {
+      self
+        .utxo_cache
+        .entry(OutPoint {
           vout: vout.try_into().unwrap(),
           txid,
-        },
-        tx_out.value,
-      );
+        })
```
Comment (raphjaph) [inline src/index/updater/inscription_updater.rs:321]: just do an insert

File: src/options.rs [hunk 15]
```diff
@@ -47,6 +47,8 @@ pub struct Options {
   pub(crate) height_limit: Option<u32>,
   #[arg(long, help = "Use index at <INDEX>.")]
   pub(crate) index: Option<PathBuf>,
+  #[arg(long, help = "Track addresses of all outputs.")]
```
Comment (raphjaph) [inline src/options.rs:50]:
```suggestion
  #[arg(long, help = "Track unspent output addresses.")]
```

File: src/re.rs [hunk 16]
```diff
@@ -5,6 +5,9 @@ fn re(s: &'static str) -> Regex {
 }
 
 lazy_static! {
+  pub(crate) static ref ADDRESS: Regex = re(
+    r"((bc1p|tb1p|bcrt1p)[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{58}|(bc1|tb1|bcrt1)[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{39,59}|[13][a-km-zA-HJ-NP-Z1-9]{25,34})"
```
Comment (raphjaph) [inline src/re.rs:9]: sanity check this somehow

File: src/subcommand/server.rs [hunk 17]
```diff
@@ -1061,7 +1091,9 @@ impl Server {
     task::block_in_place(|| {
       let query = query.trim();
 
-      if re::HASH.is_match(query) {
+      if re::ADDRESS.is_match(query) {
+        Ok(Redirect::to(&format!("/address/{query}")))
+      } else if re::HASH.is_match(query) {
```
Comment (raphjaph) [inline src/subcommand/server.rs:1096]: check address last in this regex

**Review by raphjaph — COMMENTED** [review]
Still a bunch to do

File: templates/address.html [hunk 18]
```diff
@@ -0,0 +1,7 @@
+<h1>Address {{ self.address }}</h1>
+<h2>Outputs</h2>
+<ul>
+%% for output in self.outputs.iter() {
+  <li><a href=/output/{{ output }}>{{ output }}</a></li>
```
Comment (raphjaph) [inline templates/address.html:5]:
```suggestion
  <li><a class=monospace href=/output/{{ output }}>{{ output }}</a></li>
```

File: templates/address.html [hunk 19]
```diff
@@ -0,0 +1,7 @@
+<h1>Address {{ self.address }}</h1>
+<h2>Outputs</h2>
+<ul>
+%% for output in self.outputs.iter() {
+  <li><a class=monospace href=/output/{{ output }}>{{ output }}</a></li>
+%% }
+</ul>
```
Comment (elocremarc) [inline templates/address.html:7]: This is useful for a backend like the API response. But as a front end user can we maybe paginate something that previews what actually is on the output? Like runes, inscriptions and BTC balance.

Comment (casey) [reply templates/address.html:7]: We're just doing outputs initially to keep the PR simple. Follow-up PRs can add btc/inscriptions/runes.

File: src/index.rs [hunk 20]
```diff
@@ -2186,12 +2214,26 @@ impl Index {
...
-    let sat_ranges = self.list(outpoint)?;
+  pub(crate) fn get_address_info(&self, address: &Address) -> Result<Vec<OutPoint>> {
+    self
+      .database
+      .begin_read()?
+      .open_multimap_table(SCRIPT_PUBKEY_TO_OUTPOINT)?
+      .get(address.script_pubkey().as_bytes())?
+      .map(|result| {
+        result
+          .map_err(|err| anyhow!(err.to_string()))
```
Comment (casey) [inline src/index.rs:2224]:
```suggestion
          .map_err(|err| anyhow!(err))
```

**Review by casey — APPROVED** [review]
LGTM

**Pull request merged**